Extensions/GraphQL.md

```diff
@@ -0,0 +1,37 @@
+# GraphQL Extension Spec
+
+_This extension specification is currently incubating.  While incubating the version is 0._
+
+## Introduction
+
+GraphQL protocol is one of the popular API protocol for building modern edge-network communication. This extension specification provides an interoperable structure for metadadata payloads and sub-specification which defines graphql communication over RSocket protocol.
+
+## Handling GraphQL operation
+
+A GraphQL operation can be handled if it is agreed during the RSocket `SETUP` phase and specified in the data-mime type `application/graphql` or if composite metadata enabled, specified in the REQUEST PAYLOAD data mime-type `application/graphql` via [Stream Data MIME Types Metadata Extension](/PerStreamDataMimeTypesDefinition.md).
+
+GraphQL operation can be requested via `REQUEST_RESPONSE` and `REQUEST_STREAM` requests of RSocket Protocol.
+
+> Note: +type semantic may specify graphQL message encoding if needed (e.g. `application/graphql+cbor`). By default it is JSON encoding.
+
+### Single result operation
+
+To handle `query` or `mutation` GraphQL operation a `REQUESTER` MUST issue a `REQUEST_RESPONSE` payload.
+
+GraphQL operation lifecycle is bound to the RSocket REQUEST RESPONSE operation semantic and must follow it as specified by RSocket protocol.
+
+### Streaming operation
+To handle `subscription` GraphQL operation a `REQUESTER` MUST issue a `REQUEST_STREAM` payload.
+
+Streaming operation lifecycle is bound to the RSocket REQUEST STREAM operation semantic and must follow it as specified by RSocket protocol.
+
+### Handling Graphql errors
+
+GraphQL errors must be encoded as a normal RSocket `PAYLOAD` with `NEXT` semantic followed by `PAYLOAD` with `COMPLETE` flag (onNext(ERROR_MESSAGE) -> onComplete()). RSocket Protocol `ERROR` must be treated as unexpected interruption of the subscription and MUST not contain any GraphQL data in the payload.
```

Comment:
This doesn't seem right to me. 

Compare to https://graphql.org/learn/serving-over-http/#response and https://spec.graphql.org/June2018/#sec-Errors

The spec does say "Event streams may complete in response to an error or simply because no more events will occur." so I'm happy if you know it's correct.

Isn't it valid that a stream has different errors per response, and continues?

Or is this a more serious error like a 501 in HTTP? Maybe this just needs some distinction like request level errors?

Reply:
There is a distinction between a "request error" and a "partial response".

A request error happens before a GraphQL request is accepted for execution by the GraphQL engine (e.g. parse error, query syntax error, security error, etc.) in which case the whole request fails. I think this should result in an ERROR frame.

Once the GraphQL engine starts execution, the response is considered valid but potentially partial. That is, there may be failures for individual fields (e.g. couldn't fetch the user's LinkedIn profile) which is reflected with a null value + a field specific error in the response. This is called a partial response. A subscription may have such partial responses but those do not mean the subscription ends.